pkg/storegateway/indexheader/index/postings.go

```diff
@@ -57,93 +57,139 @@ func newV1PostingOffsetTable(factory *streamencoding.DecbufFactory, tableOffset
-		t.postings[lastKey[0]][lastKey[1]] = prevRng
+		t.postings[lastKey][lastValue] = prevRng
 	}
 
 	return &t, nil
 }
 
-func newV2PostingOffsetTable(factory *streamencoding.DecbufFactory, tableOffset int, indexLastPostingEnd uint64, postingOffsetsInMemSampling int) (*PostingOffsetTableV2, error) {
+func newV2PostingOffsetTable(factory *streamencoding.DecbufFactory, tableOffset int, indexLastPostingEnd uint64, postingOffsetsInMemSampling int) (table *PostingOffsetTableV2, err error) {
 	t := PostingOffsetTableV2{
 		factory:                     factory,
 		tableOffset:                 tableOffset,
 		postings:                    map[string]*postingValueOffsets{},
 		postingOffsetsInMemSampling: postingOffsetsInMemSampling,
 	}
 
-	lastTableOff := 0
-	valueCount := 0
-	var lastKey []string
+	d := factory.NewDecbufAtChecked(tableOffset, castagnoliTable)
+	defer runutil.CloseWithErrCapture(&err, &d, "read offset table")
+
+	startLen := d.Len()
+	remainingCount := d.Be32()
+	currentName := ""
+	valuesForCurrentKey := 0
+	lastEntryOffsetInTable := -1
+
+	for d.Err() == nil && remainingCount > 0 {
+		lastName := currentName
+		offsetInTable := startLen - d.Len()
+		keyCount := d.Uvarint()
 
-	// For the postings offset table we keep every label name but only every nth
-	// label value (plus the first and last one), to save memory.
-	if err := readOffsetTable(factory, tableOffset, func(key []string, off uint64, tableOff int) error {
-		if len(key) != 2 {
-			return errors.Errorf("unexpected key length for posting table %d", len(key))
+		// The Postings offset table takes only 2 keys per entry (name and value of label).
+		if keyCount != 2 {
+			return nil, errors.Errorf("unexpected key length for posting table %d", keyCount)
 		}
 
-		if _, ok := t.postings[key[0]]; !ok {
-			// Not seen before label name.
-			t.postings[key[0]] = &postingValueOffsets{}
-			if lastKey != nil {
-				// Always include last value for each label name, unless it was just added in previous iteration based
-				// on valueCount.
-				if (valueCount-1)%postingOffsetsInMemSampling != 0 {
-					t.postings[lastKey[0]].offsets = append(t.postings[lastKey[0]].offsets, postingOffset{value: lastKey[1], tableOff: lastTableOff})
-				}
-				t.postings[lastKey[0]].lastValOffset = int64(off - crc32.Size)
-				lastKey = nil
+		// Important: this value is only valid as long as we don't perform any further reads from d.
+		// If we need to retain its value, we must copy it before performing another read.
+		unsafeName := d.UnsafeUvarintBytes()
+
+		if len(t.postings) == 0 || currentName != string(unsafeName) {
+			newKey := string(unsafeName)
+
+			if lastEntryOffsetInTable != -1 {
+				// We haven't recorded the last offset for the last value of the previous name.
+				// Go back and read the last value for the previous name.
+				newValueOffset := d.Len()
+				d.ResetAt(lastEntryOffsetInTable + 4) // 4 bytes for entry count
```

Comment:
Nit: I would define a `const entryCountLength = 4` and use it here and below as:

 ```suggestion
				d.ResetAt(entryCountLength + lastEntryOffsetInTable)
```

So you wouldn't need the comment.